### PR TITLE
Composable Balancer Stable impl

### PR DESCRIPTION
**Pull request by datluongductuan**

### CODE REVIEW

File: pkg/source/balancer-composable-stable/abis.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
+package balancer_composable_stable
```
[the remark this answers is not in the document]
Comment (datluongductuan) [reply pkg/source/balancer-composable-stable/abis.go:1]: Fixed

File: pkg/source/balancer-composable-stable/composablestable/pool_simulator.go [hunk 2]
```diff
@@ -0,0 +1,603 @@
+package composablestable
```
[the remark this answers is not in the document]
Comment (datluongductuan) [reply pkg/source/balancer-composable-stable/composablestable/pool_simulator.go:1]: fixed

File: pkg/source/balancer-composable-stable/composablestable/pool_simulator.go [hunk 3]
```diff
@@ -0,0 +1,603 @@
...
+	"strings"
+
+	"github.com/KyberNetwork/kyberswap-dex-lib/pkg/entity"
+	composableStable "github.com/KyberNetwork/kyberswap-dex-lib/pkg/source/balancer-composable-stable"
+	"github.com/KyberNetwork/kyberswap-dex-lib/pkg/source/pool"
+	"github.com/KyberNetwork/kyberswap-dex-lib/pkg/util/bignumber"
+	"github.com/KyberNetwork/kyberswap-dex-lib/pkg/valueobject"
+)
+
+type Pool struct {
```
[the remark this answers is not in the document]
Comment (datluongductuan) [reply pkg/source/balancer-composable-stable/composablestable/pool_simulator.go:15]: Fixed

File: pkg/source/balancer-composable-stable/composablestable/pool_simulator.go [hunk 4]
```diff
@@ -0,0 +1,603 @@
...
+		TotalSupply:                         totalSupply,
+		RateProviders:                       extra.RateProviders,
+		TokensExemptFromYieldProtocolFee:    extra.TokensExemptFromYieldProtocolFee,
+		TokenRateCaches:                     extra.TokenRateCaches,
+		ProtocolFeePercentageCacheSwapType:  extra.ProtocolFeePercentageCacheSwapType,
+		ProtocolFeePercentageCacheYieldType: extra.ProtocolFeePercentageCacheYieldType,
+	}, nil
+}
+
+func (c Pool) CalcAmountOut(tokenAmountIn pool.TokenAmount, tokenOut string) (*pool.CalcAmountOutResult, error) {
```
[the remark this answers is not in the document]
Comment (datluongductuan) [reply pkg/source/balancer-composable-stable/composablestable/pool_simulator.go:82]: fixed

File: pkg/source/balancer-composable-stable/composablestable/pool_simulator.go [hunk 5]
```diff
@@ -0,0 +1,603 @@
...
+	var (
+		indexIn   int
+		indexOut  int
+		amountOut *big.Int
+		fee       *pool.TokenAmount
+	)
+	//bptIndex = c.BptIndex.Int64()
+	tokens := c.Pool.GetTokens()
+	for i, token := range tokens {
```
[the remark this answers is not in the document]
Comment (datluongductuan) [reply pkg/source/balancer-composable-stable/composablestable/pool_simulator.go:91]: fixed

File: pkg/source/balancer-composable-stable/composablestable/pool_simulator.go [hunk 6]
```diff
@@ -0,0 +1,603 @@
...
+
+func (c Pool) GetMetaInfo(tokenIn string, tokenOut string) interface{} {
+	return composableStable.Meta{
+		VaultAddress: c.VaultAddress,
+		PoolId:       c.PoolId,
+	}
+}
+
+// calcBptOutGivenExactTokensIn https://github.com/balancer/balancer-v2-monorepo/blob/b46023f7c5deefaf58a0a42559a36df420e1639f/pkg/pool-stable/contracts/StableMath.sol#L201
+func calcBptOutGivenExactTokensIn(amp *big.Int, balances []*big.Int, amountsIn []*big.Int, bptTotalSupply, invariant, swapFeePercentage *big.Int) (*big.Int, *big.Int) {
```
[the remark this answers is not in the document]
Comment (datluongductuan) [reply pkg/source/balancer-composable-stable/composablestable/pool_simulator.go:135]: fixed

File: pkg/source/balancer-composable-stable/math.go [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+func init() {
+	One = new(big.Int).Set(bignumber.BONE)
+}
+
+func Upscale(amount *big.Int, scalingFactor *big.Int) *big.Int {
+	return MulDownFixed(amount, scalingFactor)
+}
+
+func ComputeScalingFactor(tokenDecimals uint) *big.Int {
+	var decimalsDiff = 36 - tokenDecimals
```
[the remark this answers is not in the document]
Comment (datluongductuan) [reply pkg/source/balancer-composable-stable/math.go:21]: this function is unused, so I will remove this

File: pkg/source/balancer-composable-stable/pools_list_updater.go [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+			weight, _ := strconv.ParseFloat(item.Weight, 64)
+			poolToken := entity.PoolToken{
+				Address:   item.Address,
+				Weight:    uint(weight * 1e18),
+				Swappable: true,
+			}
+
+			staticField.TokenDecimals = append(staticField.TokenDecimals, item.Decimals)
+			if poolToken.Weight == 0 {
+				poolToken.Weight = uint(1e18 / len(p.Tokens))
```
[the remark this answers is not in the document]
Comment (datluongductuan) [reply pkg/source/balancer-composable-stable/pools_list_updater.go:150]: I inherit this logic from the origin pools_list_updater balancer. However, calcAmountOut function doesn't use `weight` to calculate. This balancer curve math have not weighted logic. So I will remove this 

Besides, I have checked return data from graphnode, weight token for composable stable pool is always null.
`  {
        "id": "0x00c2a4be503869fa751c2dbcb7156cc970b5a8da000000000000000000000477",
        "address": "0x00c2a4be503869fa751c2dbcb7156cc970b5a8da",
        "poolType": "ComposableStable",
        "poolTypeVersion": 3,
        "tokens": [
          {
            "weight": null,
            "address": "0x00c2a4be503869fa751c2dbcb7156cc970b5a8da",
            "decimals": 18
          },
          {
            "weight": null,
            "address": "0xd4e7c1f3da1144c9e2cfd1b015eda7652b4a4399",
            "decimals": 18
          },
          {
            "weight": null,
            "address": "0xf71d0774b214c4cf51e33eb3d30ef98132e4dbaa",
            "decimals": 18
          }
        ]
      }`

File: pkg/source/balancer-composable-stable/pools_list_updater.go [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+		var staticField = StaticExtra{
+			VaultAddress: strings.ToLower(vaultAddresses[i].Hex()),
+			PoolId:       p.ID,
+		}
+
+		for _, item := range p.Tokens {
+			weight, _ := strconv.ParseFloat(item.Weight, 64)
+			poolToken := entity.PoolToken{
+				Address:   item.Address,
+				Weight:    uint(weight * 1e18),
```
[the remark this answers is not in the document]
Comment (datluongductuan) [reply pkg/source/balancer-composable-stable/pools_list_updater.go:144]: removed this logic as dicussed below

File: pkg/source/balancer-composable-stable/composablestable/pool_simulator.go [hunk 10]
```diff
@@ -0,0 +1,603 @@
...
+		if token == tokenAmountIn.Token {
+			indexIn = i
+		}
+		if token == tokenOut {
+			indexOut = i
+		}
+	}
+
+	if tokenAmountIn.Token == c.Info.Address || tokenOut == c.Info.Address {
+		amountOut, fee = c._swapWithBptGivenIn(tokenAmountIn.Amount, c.Info.Reserves, indexIn, indexOut)
```
[the remark this answers is not in the document]
Comment (datluongductuan) [reply pkg/source/balancer-composable-stable/composablestable/pool_simulator.go]: the main purpose I want to keep function signatures as the nearest same as contract logic, however, I refactored and omitted,

**Review by hanguyenk — APPROVED** [review]

**Pull request merged**